core/Network/TLS/Handshake/Server.hs

```diff
@@ -315,9 +315,32 @@ doHandshake sparams mcred ctx chosenVersion usedCipher usedCompression clientSes
             usingHState ctx $ setDHPrivate priv
             return serverParams
 
+        -- Choosing a hash algorithm to sign (EC)DHE parameters
+        -- in ServerKeyExchange. Hash algorithm is not suggested by
+        -- the chosen cipher suite. So, it should be selected based on
+        -- the "signature_algorithms" extension in a client hello.
+        -- If RSA is also used for key exchange, this function is
+        -- not called.
```

Comment:
It is not entirely true that the hash algorithm is "not suggested" by the chosen ciphersuite.  For example, a server choosing ECDSA has an ECDSA certificate with a given curve.  That curve should be one that is listed among the common hash/sig pairs between client and server.  Otherwise, if possible, the server should pick some other certificate (RSA or another EC cert if it has more than one). Such choices influence the choice of cipher:

RFC5246  page 52:
```
   If the client has offered the "signature_algorithms" extension, the
   signature algorithm and hash algorithm MUST be a pair listed in that
   extension.  Note that there is a possibility for inconsistencies
   here.  For instance, the client might offer DHE_DSS key exchange but
   omit any DSA pairs from its "signature_algorithms" extension.  In
   order to negotiate correctly, the server MUST check any candidate
   cipher suites against the "signature_algorithms" extension before
   selecting them.  This is somewhat inelegant but is a compromise
   designed to minimize changes to the original cipher suite design.
```